python/cudf/cudf/core/buffer/cow_buffer.py

```diff
@@ -0,0 +1,170 @@
+        weakref.finalize(self, _keys_cleanup, self._ptr)
+
+    @classmethod
+    def _from_device_memory(
+        cls: Type[T], data: Any, *, exposed: bool = False
+    ) -> T:
+        """Create a Buffer from an object exposing `__cuda_array_interface__`.
+
+        No data is being copied.
+
+        Parameters
+        ----------
+        data : device-buffer-like
+            An object implementing the CUDA Array Interface.
+        exposed : bool, optional
+            Mark the buffer as zero copied.
+
+        Returns
+        -------
+        Buffer
+            Buffer representing the same device memory as `data`
+        """
+
+        # Bypass `__init__` and initialize attributes manually
+        ret = super()._from_device_memory(data)
+        ret._finalize_init()
+        ret._zero_copied = exposed
+        return ret
+
+    @classmethod
+    def _from_host_memory(cls: Type[T], data: Any) -> T:
+        ret = super()._from_host_memory(data)
+        ret._finalize_init()
+        return ret
+
+    @property
+    def _is_shared(self):
+        """
+        Return `True` if `self`'s memory is shared with other columns.
+        """
+        return len(self._instances) > 1
+
+    def get_ptr(self, mode: str = "write") -> int:
+        """Device pointer to the start of the buffer.
+
+        Parameters
+        ----------
+        mode : str, default 'write'
+            Supported values are {"read", "write"}
+            If "write", when weak-references exist, they
+            are unlinked and the data pointed to may be modified
+            by the caller. If "read", the data pointed to
+            must not be modified by the caller.
+            Failure to fulfill this contract will cause
+            incorrect behavior.
+
+        See Also
+        --------
+        Buffer.get_ptr
+        SpillableBuffer.get_ptr
+        """
+        if mode == "write":
+            self._unlink_shared_buffers()
+        elif mode != "read":
+            raise ValueError(f"Incorrect mode passed : {mode}")
+        return self._ptr
+
+    def copy(self, deep: bool = True):
+        if deep or self._zero_copied:
+            return super().copy(deep=True)
+        else:
+            cls = type(self)
+            copied_buf = cls.__new__(cls)
+            copied_buf._ptr = self._ptr
+            copied_buf._size = self._size
+            copied_buf._owner = self._owner
+            copied_buf._finalize_init()
+            return copied_buf
+
+    @property
+    def __cuda_array_interface__(self) -> dict:
+        self._unlink_shared_buffers()
+        # Unlink if there are any weak references.
+        # Mark the Buffer as ``zero_copied=True``,
+        # which will prevent any copy-on-write
+        # mechanism post this operation.
+        # This is done because we don't have any
+        # control over knowing if a third-party library
+        # has modified the data this Buffer is
+        # pointing to.
+        self._zero_copied = True
```

Comment:
```suggestion
        # Unlink if there are any weak references.
        # Mark the Buffer as ``zero_copied=True``,
        # which will prevent any copy-on-write
        # mechanism post this operation.
        # This is done because we don't have any
        # control over knowing if a third-party library
        # has modified the data this Buffer is
        # pointing to.
        self._unlink_shared_buffers()
        self._zero_copied = True
```